miniapps/meshing/minimal-surface.cpp

```diff
@@ -0,0 +1,1302 @@
+      const int print_iter = -1, max_num_iter = 2000;
+      const double RTOLERANCE = EPS, ATOLERANCE = EPS*EPS;
+   public:
+      Solver(Surface &S, Opt &opt): opt(opt), S(S), cg(),
+         a(S.fes), x(S.fes), x0(S.fes), b(S.fes), one(1.0)
+      {
+         cg.SetRelTol(RTOLERANCE);
+         cg.SetAbsTol(ATOLERANCE);
+         cg.SetMaxIter(max_num_iter);
+         cg.SetPrintLevel(print_iter);
+      }
+
+      ~Solver() { delete M; }
+
+      void Solve()
+      {
+         constexpr bool converged = true;
+         if (opt.pa) { a.SetAssemblyLevel(AssemblyLevel::PARTIAL);}
+         for (int i=0; i < opt.niters; ++i)
+         {
+            if (opt.amr) { Amr(); }
+            if (opt.vis) { Surface::Visualize(opt, S.mesh); }
+            if (!opt.id) { mfem::out << "Iteration " << i << ": "; }
+            S.mesh->DeleteGeometricFactors();
+            a.Update();
+            a.Assemble();
+            if (Step() == converged) { break; }
+         }
+         if (opt.print) { Surface::Print(opt, S.mesh, S.mesh->GetNodes()); }
+      }
+
+      virtual bool Step() = 0;
+
+   protected:
+      bool Converged(const double rnorm) { return rnorm < NRM; }
+
+      bool ParAXeqB()
+      {
+         b = 0.0;
+         Vector X, B;
+         a.FormLinearSystem(S.bc, x, b, A, X, B);
+         if (!opt.pa) { M = new GSSmoother((SparseMatrix&)(*A)); }
+         if (M) { cg.SetPreconditioner(*M); }
+         cg.SetOperator(*A);
+         cg.Mult(B, X);
+         a.RecoverFEMSolution(X, b, x);
+         const bool by_vdim = opt.by_vdim;
+         GridFunction *nodes = by_vdim ? &x0 : S.fes->GetMesh()->GetNodes();
+         x.HostReadWrite();
+         nodes->HostRead();
+         double rnorm = nodes->DistanceTo(x) / nodes->Norml2();
+         if (!opt.id) { mfem::out << "rnorm = " << rnorm << std::endl; }
+         const double lambda = opt.lambda;
+         if (by_vdim)
+         {
+            MFEM_VERIFY(!opt.radial,"'VDim solver can't use radial option!");
+            return Converged(rnorm);
+         }
+         if (opt.radial)
+         {
+            GridFunction delta(S.fes);
+            subtract(x, *nodes, delta); // Δ = x - nodes
+            // position and Δ vectors at point i
+            Vector ni(SDIM), di(SDIM);
+            for (int i = 0; i < delta.Size()/SDIM; i++)
+            {
+               // extract local vectors
+               const int ndof = S.fes->GetNDofs();
+               for (int d = 0; d < SDIM; d++)
+               {
+                  ni(d) = (*nodes)(d*ndof + i);
+                  di(d) = delta(d*ndof + i);
+               }
+               // project the delta vector in radial direction
+               const double ndotd = (ni*di) / (ni*ni);
+               di.Set(ndotd,ni);
+               // set global vectors
+               for (int d = 0; d < SDIM; d++) { delta(d*ndof + i) = di(d); }
+            }
+            add(*nodes, delta, *nodes);
+         }
+         // x = λ*nodes + (1-λ)*x
```

Comment:
Do we really need these special characters (lambda, delta above) in comments?

Reply:
I removed them, right.

Reply:
`done`